Python/ceval.c

```diff
@@ -2903,6 +2903,19 @@ _PyEval_ImportFrom(PyThreadState *tstate, PyObject *v, PyObject *name)
         }
         else {
             assert(rc == 0);
+            if (origin == NULL) {
+                // Fall back to __file__ for diagnostics if we don't have
+                // an origin that is a location
+                origin = PyModule_GetFilenameObject(v);
+                if (origin == NULL) {
+                    if (!PyErr_ExceptionMatches(PyExc_SystemError)) {
```

Comment:
In what cases it fails? Is there a test for this?

Reply:
Thanks, I added two test cases.

`PyModule_GetFilenameObject` can set system error for a few reasons, such as `__file__` attribute missing or `__file__` not being a str.

I'm not sure when exactly you'd have an error set that is not system error (although `PyModule_GetFilenameObject` does check `PyErr_Occurred` before setting system error). This code is based on what it was prior to #123929, see the left hand side of: https://github.com/python/cpython/pull/123929/files#diff-c22186367cbe20233e843261998dc027ae5f1f8c0d2e778abfa454ae74cc59deL2840-L2849

Reply:
We only use `origin` to improve error messages. I think it would be strange for a user to get `SystemError: module filename missing` when doing `import os; del os.__file__; from os import does_not_exist`

Reply:
This code originates from this commit of yours: https://github.com/python/cpython/commit/8660fb7fd7cd#diff-c22186367cbe20233e843261998dc027ae5f1f8c0d2e778abfa454ae74cc59deR2654

Reply:
Ah, right. I misread the code. The error is cleared only if it _is_ a SystemError. It is okay.